text/0056-sequence-number-panorama.md

```diff
@@ -0,0 +1,153 @@
+# Sequence number panoramas
+
+## Summary
+
+[summary]: #summary
+
+CEP PR: [casperlabs/ceps#0056](https://github.com/casperlabs/ceps/pull/0056)
+
+We propose citing units by sequence number instead of by hash.
+This will reduce the protocol overhead and bandwidth requirement.
+
+
+## Motivation
+
+[motivation]: #motivation
+
+32-byte hashes of earlier units make up the largest part of Highway messages.
+Replacing them by 8-byte sequence numbers will therefore considerably reduce message sizes,
+and improve performance.
+
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Each unit — the main type of message in the Highway protocol — contains a list of hashes of earlier units.
+That way the sender attests to which units they had received before they created the new one.
+This is what defines the directed acyclic graph (DAG) structure of Highway's protocol state:
+The graph has an arrow `u → v` if `u` _directly cites_ `v`, i.e. the message `u` mentions `v` by hash.
```

Comment:
```suggestion
The graph has an arrow `u → v` if `u` _directly cites_ `v`, i.e. the hash of `v` was used to generate `u`'s hash.
```

Reply:
But it's not really about `u`'s hash, is it? It's about whether `u` attests to `v`, by mentioning `v` by hash. (Currently — or by sequence number and control hash after this CEP.)

Reply:
Hmm... But it might be a good point - the hash of a unit should still be calculated using the hash-based panorama, right? Otherwise we lose the guarantee that the hash of a unit uniquely identifies both a unit _and_ its ancestors. Is it still calculated this way in your PR? I remember you changed some fields in the unit struct there.

Maybe this is actually worth stressing somewhere in the CEP? Not necessarily here, though, because I agree that the point raised here isn't about hashes, it's more of an expression of an abstract property.

Reply:
Ah, the unit still contains the hash of the (hash-based) panorama, so it should be fine :+1:

Reply:
Ah, I see, yes, maybe that's what @xcthulhu means?

Reply:
> the hash of a unit should still be calculated using the hash-based panorama, right?

@fizyk20 This is what I am trying to express, correct.

It's crucial to emphasize that cryptographic integrity does not rely on sequence numbers alone, but hashes of directly cited units.  A directly cited hash uses its sequence number as part of how the hash is generated.

Switching to sequence numbers is basically switching to an index.  Most of the time nodes will not be faulty.  While none of the nodes are faulty we can optimistically use indices.  When the system is faulty, we need to fall back to the slow but bulletproof hash-based approach of getting units from peers.

Reply:
OK, I'll add that to the text. I'd like to do it in a different line, though, since this one is about defining the order itself, and I'd like to separate that from the cryptographic integrity question.